test/test_autodiff.hpp

```diff
@@ -8,90 +8,108 @@
 #include <boost/multiprecision/cpp_bin_float.hpp>
 #include <boost/multiprecision/cpp_dec_float.hpp>
+
+#include <boost/mp11/integral.hpp>
+#include <boost/mp11/list.hpp>
+#include <boost/mp11/mpl.hpp>
+#include <boost/mp11/utility.hpp>
 #include <boost/range/irange.hpp>
 
+#include <algorithm>
 #include <cfenv>
 #include <cstdlib>
 #include <random>
 
 #define BOOST_TEST_MODULE test_autodiff
 #include <boost/test/included/unit_test.hpp>
+#include <boost/test/tools/floating_point_comparison.hpp>
 
 #if defined(_MSC_VER) || defined(BOOST_MSVC)
 #define NOMINMAX
 #endif
 
+namespace mp11 = boost::mp11;
+namespace bmp = boost::multiprecision;
+
 // using bin_float_types = mp_list<float,double,long
-// double,boost::multiprecision::cpp_bin_float_50>;
-using bin_float_types = boost::mp11::mp_list<float, double, long double>;
+// double,bmp::cpp_bin_float_50>;
+using bin_float_types = mp11::mp_list<float, double, long double>;
 //  cpp_bin_float_50 is fixed in boost 1.70
 // float blows up in unchecked_factorial
 
 // cpp_dec_float_50 cannot be used with close_at_tolerance
 /*using multiprecision_float_types =
-    mp_list<boost::multiprecision::cpp_dec_float_50, boost::multiprecision::cpp_bin_float_50>;*/
-using multiprecision_float_types = boost::mp11::mp_list<>;
+    mp_list<bmp::cpp_dec_float_50, bmp::cpp_bin_float_50>;*/
+using multiprecision_float_types = mp11::mp_list<>;
 
-using all_float_types = boost::mp11::mp_append<bin_float_types, multiprecision_float_types>;
+using all_float_types = mp11::mp_append<bin_float_types, multiprecision_float_types>;
 
 using namespace boost::math::differentiation;
 
+template <typename... Ts> struct print_vals;
+
 namespace test_detail {
 
 /**
  * struct to emit pseudo-random values from a given interval.
  * Endpoints are closed or open depending on whether or not they're infinite).
  */
-template <typename T, typename = void>
-struct RandomSample;
-
-template <typename T>
-struct RandomSample<
-    T, typename std::enable_if<std::is_floating_point<T>::value || std::numeric_limits<T>::is_integer>::type> {
-  using dist_t = typename boost::conditional<std::is_floating_point<T>::value, std::uniform_real_distribution<T>,
-                                             std::uniform_int_distribution<T>>::type;
+
+template <typename T> struct RandomSample {
+  using is_multiprecision_t = mp11::mp_or<bmp::is_number<T>, bmp::is_number_expression<T>>;
+  using numeric_limits_t = std::numeric_limits<T>;
+  using is_integer_t = mp11::mp_bool<std::numeric_limits<T>::is_integer>;
+
+  using distribution_param_t =
+      mp11::mp_if<is_multiprecision_t,
+                  mp11::mp_if<is_integer_t, mp11::mp_if_c<numeric_limits_t::is_signed, int64_t, uint64_t>, long double>,
+                  T>;
+  static_assert(std::numeric_limits<T>::is_integer || !std::numeric_limits<distribution_param_t>::is_integer,
+                "T and distribution_param_t must either both be integral or both be not integral");
```

Comment:
This code doesn't seem to match the message. To me the message would imply
``` std::numeric_limits<T>::is_integer == std::numeric_limits<distribution_param_t>::is_integer ```

Reply:
fixed